Firestore/core/src/local/leveldb_key.h

```diff
@@ -1145,6 +1153,73 @@ class LevelDbDocumentOverlayCollectionIndexKey
   }
 };
 
+/** A key in the "collection group" index of the document_overlays table. */
+class LevelDbDocumentOverlayCollectionGroupIndexKey
+    : public LevelDbDocumentOverlayIndexKey {
+ public:
+  /**
+   * Creates a key prefix that points just before the first key for the given
+   * user_id.
+   */
+  static std::string KeyPrefix(absl::string_view user_id);
+
+  /**
+   * Creates a key prefix that points just before the first key for the given
+   * user_id, collection group.
+   */
+  static std::string KeyPrefix(absl::string_view user_id,
+                               absl::string_view collection_group);
+
+  /**
+   * Creates a key prefix that points just before the first key for the given
+   * user_id, collection group, and largest_batch_id.
+   */
+  static std::string KeyPrefix(absl::string_view user_id,
+                               absl::string_view collection_group,
+                               model::BatchId largest_batch_id);
+
+  /**
+   * Creates a complete key that points to a specific user_id, collection group,
+   * and largest_batch_id for a given key in the document_overlays table.
+   */
+  static std::string Key(absl::string_view user_id,
+                         absl::string_view collection_group,
+                         model::BatchId largest_batch_id,
+                         const model::DocumentKey& document_key);
+
+  /**
+   * Creates a complete key that points to a specific key in the overlays table.
+   */
+  static absl::optional<std::string> Key(const LevelDbDocumentOverlayKey& key) {
+    const absl::optional<std::string> collection_group =
+        key.document_key().GetCollectionId();
+    if (!collection_group.has_value()) {
```

Comment:
Ideally, this should never happen. This means that we are trying to insert a `DocumentKey::Empty()`, which points to a bug. Can we assert here and return a non-optional string?

Reply:
Done. I've changed the graceful handing of a `DocumentKey` without a collection group to a `HARD_ASSERT`.